File: C/problem004.c
/*
 * Problem 4 - Project Euler
 * http://projecteuler.net/index.php?section=problems&id=004
 */
#include <math.h>
#include <stdio.h>

/*
- 999x999から始める。
- 最大のものは6桁になるだろうという前提で処理する(10000 < y < 1000000)
 */
void problem004() {
    int ans;
    int i, j, target, tmp, digit_num;
    int digit[6];

    /* solve this problem here */
    ans = 0;
    for (i = 999; i >= 101; i--) {
        for (j = 999; j >= 101; j--) {
            target = i * j;
            if (target < ans || target < 100000) {
                /* 現状の最大値より小さいまたは5桁になったらループを一つ抜ける
                 */
                break;
            }
            for (digit_num = 0; digit_num < 6; digit_num++) {
                tmp = (int)pow(10, digit_num + 1);
                digit[digit_num] = target - ((target / tmp) * tmp);
                digit[digit_num] /= (int)pow(10, digit_num);
            }
            if (digit[0] == digit[5] && digit[1] == digit[4] &&
                digit[2] == digit[3]) {
                if (target > ans) {
                    printf("(i,j) = (%d, %d), i*j = %d\n", i, j, target);
                    ans = target;
                }
            }
        }
    }

    printf("%s: answer = %d\n", __FUNCTION__, ans);
}

int main(int argc, char* argv[]) {
    problem004();
    return 0;
}

File: Zig/problem004.zig
// zig version 0.10.0
const std = @import("std");
const stdout = std.io.getStdOut().writer();
const ArrayList = std.ArrayList;
const allocator = std.heap.page_allocator;

// #4 Largest palindrome product - Project Euler
// http://projecteuler.net/index.php?section=problems&id=4
pub fn main() !void {
    var ans: u32 = 0;

    for (100..1000) |i| {
        for (i..1000) |j| {
            const x = @intCast(u32, i * j);
            if (ans < x and try is_palindromic(x)) {
                ans = x;
            }
        }
    }

    try stdout.print("{}\n", .{ans});
}

fn is_palindromic(n: u32) !bool {
    var digits = ArrayList(u32).init(allocator);
    defer digits.deinit();

    var i: usize = 0;
    var x: u32 = n;
    while (x > 0) : (x = @divTrunc(x, 10)) {
        try digits.append(@mod(x, 10));
        i += 1;
    }

    var ret = false;
    const len = digits.items.len;
    for (0..@divTrunc(len, 2)) |j| {
        if (digits.items[j] != digits.items[len - j - 1]) {
            break;
        }
    } else {
        ret = true;
    }

    return ret;
}
